### PR TITLE
feat: implement JUMP instruction

### PR DESCRIPTION
**Pull request by lambda-0x**
## Pull Request type

Please check the type of change your PR introduces:

- [x] Feature

## What is the current behavior?

Resolves: #250 

## What is the new behavior?

-
-
-

## Does this introduce a breaking change?

- [ ] Yes
- [ ] No

### CODE REVIEW

File: crates/evm/src/instructions/memory_operations.cairo [hunk 1]
```diff
@@ -49,7 +49,28 @@ impl MemoryOperation of MemoryOperationTrait {
...
     fn exec_jump(ref self: ExecutionContext) -> Result<(), EVMError> {
-        panic_with_felt252('JUMP not implement yet')
+        let idx = self.stack.pop_usize()?;
+
+        // TODO: Currently this doesn't check that byte is actually `JUMPDEST`
+        // and not `0x5B` that is a part of PUSHN instruction
+        // 
+        // That can be done by storing all valid jump locations during contract deployment
+        // which would also simplify the logic because we would be just checking if idx is
+        // present in that list
```
Comment (lambda-0x) [inline crates/evm/src/instructions/memory_operations.cairo:59]: as discussed in telegram channel, to do this we would need to store a map of valid jump locations in `Contract` which currently don't exist.

this issue will be similar for `JUMP` and `JUMPI` should i create a different tracking issue for this?

Comment (Eikix) [reply crates/evm/src/instructions/memory_operations.cairo:59]: good idea

File: crates/evm/src/instructions/memory_operations.cairo [hunk 2]
```diff
@@ -49,7 +49,28 @@ impl MemoryOperation of MemoryOperationTrait {
     /// The new pc target has to be a JUMPDEST opcode.
     /// # Specification: https://www.evm.codes/#56?fork=shanghai
     fn exec_jump(ref self: ExecutionContext) -> Result<(), EVMError> {
```
Comment (Eikix) [inline crates/evm/src/instructions/memory_operations.cairo:51]: For clarity let's add this as documentation above the exec_jump?
```
 Valid jump destinations are defined as follows:
        * The jump destination is less than the length of the code.
        * The jump destination should have the `JUMPDEST` opcode (0x5B).
        * The jump destination shouldn't be part of the data corresponding to
          `PUSH-N` opcodes.

    Note - Jump destinations are 0-indexed.
```

**Review by Eikix — COMMENTED** [review]
Nice!

**Review by enitrat — COMMENTED** [review]
Please create an issue describing the next task regarding the JUMP opcode

**Comment by lambda-0x** [comment]
Opened: https://github.com/kkrt-labs/kakarot-ssj/issues/353

**Review by Eikix — APPROVED** [review]
lgtm

File: crates/evm/src/instructions/memory_operations.cairo [hunk 3]
```diff
@@ -48,8 +48,37 @@ impl MemoryOperation of MemoryOperationTrait {
...
+    ///  Valid jump destinations are defined as follows:
+    ///     * The jump destination is less than the length of the code.
+    ///     * The jump destination should have the `JUMPDEST` opcode (0x5B).
+    ///     * The jump destination shouldn't be part of the data corresponding to
+    ///       `PUSH-N` opcodes.
+    ///
+    /// Note: Jump destinations are 0-indexed.
     fn exec_jump(ref self: ExecutionContext) -> Result<(), EVMError> {
-        panic_with_felt252('JUMP not implement yet')
+        let idx = self.stack.pop_usize()?;
```
Comment (Eikix) [inline crates/evm/src/instructions/memory_operations.cairo:60]:
```suggestion
        let index = self.stack.pop_usize()?;
```

File: crates/evm/src/instructions/memory_operations.cairo [hunk 4]
```diff
@@ -48,8 +48,37 @@ impl MemoryOperation of MemoryOperationTrait {
...
+            Option::Some(op) => {
+                if *op.unbox() != 0x5B {
+                    return Result::Err(EVMError::JumpError(INVALID_DESTINATION));
+                }
+            },
+            Option::None => {
+                return Result::Err(EVMError::JumpError(INVALID_DESTINATION));
+            }
+        }
+        self.program_counter = idx;
```
Comment (Eikix) [inline crates/evm/src/instructions/memory_operations.cairo:80]:
```suggestion
        self.program_counter = index;
```

File: crates/evm/src/instructions/memory_operations.cairo [hunk 5]
```diff
@@ -48,8 +48,37 @@ impl MemoryOperation of MemoryOperationTrait {
...
+        // TODO: Currently this doesn't check that byte is actually `JUMPDEST`
+        // and not `0x5B` that is a part of PUSHN instruction
+        // 
+        // That can be done by storing all valid jump locations during contract deployment
+        // which would also simplify the logic because we would be just checking if idx is
+        // present in that list
+        //
+        // Check if idx in bytecode points to `JUMPDEST` opcode
+        match self.call_context().bytecode.get(idx) {
+            Option::Some(op) => {
```
Comment (Eikix) [inline crates/evm/src/instructions/memory_operations.cairo:71]:
```suggestion
            Option::Some(opcode) => {
```

File: crates/evm/src/instructions/memory_operations.cairo [hunk 6]
```diff
@@ -48,8 +48,37 @@ impl MemoryOperation of MemoryOperationTrait {
...
+
+        // TODO: Currently this doesn't check that byte is actually `JUMPDEST`
+        // and not `0x5B` that is a part of PUSHN instruction
+        // 
+        // That can be done by storing all valid jump locations during contract deployment
+        // which would also simplify the logic because we would be just checking if idx is
+        // present in that list
+        //
+        // Check if idx in bytecode points to `JUMPDEST` opcode
+        match self.call_context().bytecode.get(idx) {
```
Comment (Eikix) [inline crates/evm/src/instructions/memory_operations.cairo:70]:
```suggestion
        match self.call_context().bytecode.get(index) {
```

File: crates/evm/src/instructions/memory_operations.cairo [hunk 7]
```diff
@@ -48,8 +48,37 @@ impl MemoryOperation of MemoryOperationTrait {
...
+            Option::Some(op) => {
+                if *op.unbox() != 0x5B {
```
Comment (Eikix) [inline crates/evm/src/instructions/memory_operations.cairo:72]:
```suggestion
                if *opcode.unbox() != 0x5B {
```

**Review by Eikix — COMMENTED** [review]
Let's try to keep full words as much as possible in the codebase, as discussed with @ClementWalter

**Comment by lambda-0x** [comment]
done

**Review by Eikix — APPROVED** [review]
lgtm

**Pull request merged**